pkg/controller/elasticsearch/driver/upgrade_deletion_strategy.go

```diff
@@ -0,0 +1,299 @@
+			if !label.IsMasterNode(candidate) {
+				return true, nil
+			}
+
+			// If only one master node is expected this cluster is not H.A.
+			if len(d.masterNodesNames) < 2 {
+				return true, nil
+			}
+
+			var healthyMasters []corev1.Pod
+			candidateIsHealthy := false
+			for _, expectedMaster := range d.masterNodesNames {
+				for healthyPodName, healthyPod := range d.healthyPods {
+					if !label.IsMasterNode(healthyPod) {
+						continue
+					}
+					master := types.NamespacedName{
+						Name:      expectedMaster,
+						Namespace: candidate.Namespace,
+					}
+					if candidate.Name == healthyPodName {
+						candidateIsHealthy = true
+					}
+					if healthyPodName == master.Name {
+						healthyMasters = append(healthyMasters, healthyPod)
+					}
+				}
+			}
+
+			if candidateIsHealthy && (len(healthyMasters) == 1) {
+				// Last healthy one, don't delete
+				return false, nil
+			}
+			return true, nil
+		},
+		"skip_unknown_or_long_terminating_pods": func(
+			candidate corev1.Pod,
+			expectedDeletions []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			if candidate.DeletionTimestamp != nil && candidate.Status.Reason == NodeUnreachablePodReason {
+				// kubelet is unresponsive, Unknown Pod, do not try to delete it
+				return false, nil
+			}
+			if candidate.DeletionTimestamp != nil && time.Now().After(candidate.DeletionTimestamp.Add(TerminatingPodTimeout)) {
+				return false, nil
+			}
+			return true, nil
+		},
+		// We should not delete 2 Pods with the same shards
+		"do_not_delete_pods_with_same_shards": func(
+			candidate corev1.Pod,
+			expectedDeletions []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			if len(expectedDeletions) == 0 {
+				// Do not do unnecessary request
+				return true, nil
+			}
+			clusterState, err := d.esState.GetClusterState()
+			if err != nil {
+				return false, err
+			}
+			shards := clusterState.GetShardsByNode()
+			shardsOnCandidate, ok := shards[candidate.Name]
+			if !ok {
+				// No shards on this node
+				return true, nil
+			}
+
+			for _, deletedPod := range expectedDeletions {
+				shardsOnDeletePod, ok := shards[deletedPod.Name]
+				if !ok {
+					// No shards on the deleted pod
+					continue
+				}
+				if conflict(shardsOnCandidate, shardsOnDeletePod) {
+					return false, nil
+				}
+			}
+			return true, nil
+		},
+		// In Yellow or Red status only allow unhealthy Pods to be restarted.
+		// This is intended to unlock some situations where the cluster is not green and
+		// a Pod has to be restarted a second time.
+		"do_not_restart_healthy_node_if_not_green": func(
+			candidate corev1.Pod,
+			expectedDeletions []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			green, err := d.esState.GreenHealth()
+			if err != nil {
+				return false, err
+			}
+			if green {
+				return true, nil
+			}
+			if !k8s.IsPodReady(candidate) {
+				return true, nil
+			}
+			return false, nil
+		},
+	}
+}
```

Comment:
This whole thing is getting a bit long maybe worth pulling out into separate functions. You could still have descriptions for logging if you had a `Predicate` type like
```go
type Predicate struct {
name string
fn func(....) (bool, errro)
}
```
and then have a `[]Predicate` which would have the added benefit of a guaranteed iteration order when applying the predicates

Reply:
I'm not sure about the right way to split it. Note that now it is a constant array and not a function.